File: tests/netDxf.Conformance/BinarySentinelTests.cs
using System.Text;

namespace NetDxf.Conformance;

internal static partial class Program
{
    private static void RegisterBinarySentinelTests()
    {
        for (int length = 0; length < BinarySentinel.Length; length++)
        {
            int captured = length;
            Run($"binary/sentinel/truncated-{length}", () =>
            {
                using var stream = new MemoryStream(BinarySentinel[..captured]);
                using var reader = new BinaryReader(stream, Encoding.UTF8, true);
                Throws<EndOfStreamException>(() => CreateBinaryReader(reader));
            });
        }
        for (int index = 0; index < BinarySentinel.Length; index++)
        {
            int captured = index;
            Run($"binary/sentinel/corrupt-byte-{index}", () =>
            {
                byte[] bytes = (byte[])BinarySentinel.Clone();
                bytes[captured] ^= 0x01;
                using var stream = new MemoryStream(bytes);
                using var reader = new BinaryReader(stream, Encoding.UTF8, true);
                Throws<InvalidDataException>(() => CreateBinaryReader(reader));
            });
        }
        Run("binary/sentinel/fragmented-stream", () =>
        {
            using var stream = new FragmentedReadStream(BinarySentinel);
            using var reader = new BinaryReader(stream, Encoding.UTF8, true);
            CreateBinaryReader(reader);
            Equal((long)BinarySentinel.Length, stream.Position, "sentinel consumption");
        });
        Run("binary/sentinel/null-reader", () =>
            Throws<ArgumentNullException>(() => CreateBinaryReader(null!)));
        Run("binary/sentinel/no-overread", () =>
        {
            byte[] bytes = BinarySentinel.Concat(new byte[] { 0x55, 0xAA }).ToArray();
            using var stream = new MemoryStream(bytes);
            using var reader = new BinaryReader(stream, Encoding.UTF8, true);
            CreateBinaryReader(reader);
            Equal((byte)0x55, reader.ReadByte(), "next byte after sentinel");
        });
    }

    private sealed class FragmentedReadStream : MemoryStream
    {
        internal FragmentedReadStream(byte[] bytes) : base(bytes, false) { }
        public override int Read(byte[] buffer, int offset, int count) =>
            base.Read(buffer, offset, Math.Min(count, 1));
        public override int Read(Span<byte> buffer) =>
            base.Read(buffer[..Math.Min(buffer.Length, 1)]);
    }
}

File: javascript/tests/netDxf.Conformance/BinarySentinelTests.js
// Complete original BinarySentinelTests.cs; the BinaryReader wrapper is the existing stream codec adapter.
import { MemoryStream } from '../../runtime/MemoryStream.js';
import { BinarySentinel } from '../../runtime/BinaryCursor.js';
import { BinaryCodeValueReader } from '../../netDxf/IO/BinaryCodeValueReader.js';
import { ArgumentNullException, EndOfStreamException, InvalidDataException } from '../../runtime/Errors.js';
import { Run, Equal, Throws } from './TestHarness.js';
export class FragmentedReadStream extends MemoryStream {
  constructor(bytes){super(bytes,false);}
  Read(buffer,offset,count){return super.Read(buffer,offset,Math.min(count,1));}
}
export function RegisterBinarySentinelTests(){
  for(let length=0;length<BinarySentinel.length;length++)Run(`binary/sentinel/truncated-${length}`,()=>{
    const stream=new MemoryStream(BinarySentinel.slice(0,length));try{Throws(EndOfStreamException,()=>new BinaryCodeValueReader(stream));}finally{stream.Dispose();}
  });
  for(let index=0;index<BinarySentinel.length;index++)Run(`binary/sentinel/corrupt-byte-${index}`,()=>{
    const bytes=BinarySentinel.slice();bytes[index]^=1;const stream=new MemoryStream(bytes);try{Throws(InvalidDataException,()=>new BinaryCodeValueReader(stream));}finally{stream.Dispose();}
  });
  Run('binary/sentinel/fragmented-stream',()=>{const stream=new FragmentedReadStream(BinarySentinel.slice());try{new BinaryCodeValueReader(stream);Equal(BinarySentinel.length,stream.Position,'sentinel consumption');}finally{stream.Dispose();}});
  Run('binary/sentinel/null-reader',()=>Throws(ArgumentNullException,()=>new BinaryCodeValueReader(null)));
  Run('binary/sentinel/no-overread',()=>{const stream=new MemoryStream(Uint8Array.from([...BinarySentinel,0x55,0xaa]));try{new BinaryCodeValueReader(stream);Equal(0x55,stream.ReadByte(),'next byte after sentinel');}finally{stream.Dispose();}});
}
